### PR TITLE
Thread param

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This was accumulated when preparing the SPCI prototype, but makes sense to introduce by itself as it improves the current code.

There's an abstraction layer introduced for RPC. Compared to current state there is no added overhead, comparing the size of generated code there's actually a small reduction in size.

### CODE REVIEW

File: core/arch/arm/include/kernel/thread.h [hunk 1]
```diff
@@ -595,15 +594,58 @@ struct mobj *thread_rpc_alloc_payload(size_t size);
...
+};
+
+enum thread_param_attr {
+	THREAD_PARAM_ATTR_NONE = 0,
+	THREAD_PARAM_ATTR_VALUE_IN,
+	THREAD_PARAM_ATTR_VALUE_OUT,
+	THREAD_PARAM_ATTR_VALUE_INOUT,
+	THREAD_PARAM_ATTR_MEMREF_IN,
+	THREAD_PARAM_ATTR_MEMREF_OUT,
+	THREAD_PARAM_ATTR_MEMREF_INOUT,
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/include/kernel/thread.h]: I'll add a comment

File: core/tee/tee_fs_rpc.c [hunk 2]
```diff
@@ -47,11 +39,11 @@ static TEE_Result operation_open(uint32_t id, unsigned int cmd,
 	if (!va)
 		return TEE_ERROR_OUT_OF_MEMORY;
 
-	op.params[0].attr = THREAD_PARAM_ATTR_VALUE_IN;
-	op.params[0].u.value.a = cmd;
-
-	init_memparam(op.params + 1, mobj, 0, TEE_FS_NAME_MAX,
-		      THREAD_PARAM_ATTR_MEMREF_IN);
+	struct tee_fs_rpc_operation op = {
+		.id = id, .num_params = 3, .params = {
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/tee/tee_fs_rpc.c]: Yes, it looks suspicious. The code is actually correct, but the entire `op` should be initialized at once, I'll clean up.

File: core/arch/arm/tee/pta_socket.c [hunk 3]
```diff
@@ -159,29 +132,26 @@ static TEE_Result socket_recv(uint32_t instance_id, uint32_t param_types,
...
 
-	msg_params[2].attr = THREAD_PARAM_ATTR_VALUE_IN;
-	msg_params[2].u.value.a = params[0].value.b /* timeout */;
+	res = thread_rpc_cmd(OPTEE_MSG_RPC_CMD_SOCKET, 3, tpm);
 
-
-	res = thread_rpc_cmd(OPTEE_MSG_RPC_CMD_SOCKET, 3, msg_params);
-	params[1].memref.size = msg_params[1].u.memref.size;
+	if (tpm[1].u.memref.size > params[1].memref.size)
+		return TEE_ERROR_GENERIC;
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/tee/pta_socket.c]: It's possible to design a protocol like that, but that's not how it's done here (this uses the `read(2)` style of call). The test is just a safety check.

**Comment by jenswi-linaro** [comment]
Comments addressed, tags applied.

**Comment by jforissier** [comment]
The checkpatch error is easy to fix (spelling error).

**Comment by jenswi-linaro** [comment]
Sorry about the checkpatch error, fixed now.

File: core/arch/arm/kernel/thread.c [hunk 4]
```diff
@@ -1411,26 +1412,69 @@ static struct mobj *thread_rpc_alloc_arg(size_t size)
...
+	}
+
+	param->u.tmem.size = tpm->u.memref.size;
+	param->u.tmem.buf_ptr = pa;
+	param->u.tmem.shm_ref = shm_ref;
+
+	return true;
+}
+
+static uint32_t get_rpc_arg(uint32_t cmd, size_t num_params,
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread.c]: Should we replace return type `uint32_t` with `TEE_Result`?

File: core/arch/arm/kernel/thread.c [hunk 5]
```diff
@@ -1442,45 +1486,95 @@ static bool get_rpc_arg(uint32_t cmd, size_t num_params,
 	arg->num_params = num_params;
 	arg->ret = TEE_ERROR_GENERIC; /* in case value isn't updated */
 
+	for (size_t n = 0; n < num_params; n++) {
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread.c]: Yes, we decided to use that style a month ago or so.

File: core/arch/arm/kernel/thread.c (hunk 4 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread.c]: Sure, but I think it's better to do that after this PR.

Comment (etienne-lms) [reply core/arch/arm/kernel/thread.c]: ok

File: core/arch/arm/kernel/thread.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/arch/arm/kernel/thread.c]: ok

File: core/arch/arm/kernel/wait_queue.c [hunk 6]
```diff
@@ -6,7 +6,7 @@
 #include <types_ext.h>
 #include <tee_api_defines.h>
 #include <string.h>
-#include <optee_msg.h>
+#include <optee_rpc_cmd.h>
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/wait_queue.c]: I'll sort the includes

File: core/arch/arm/kernel/thread.c [hunk 7]
```diff
@@ -1574,24 +1639,52 @@ static struct mobj *thread_rpc_alloc(size_t size, size_t align, unsigned int bt)
...
 
+/**
+ * Allocates shared memory buffer via RPC
+ *
+ * @size:	size in bytes of shared memory buffer
+ * @align:	required alignment of buffer
+ * @bt:		buffer type OPTEE_RPC_SHM_TYPE_*
+ * @payload:	returned physical pointer to buffer, 0 if allocation
+ *		failed.
+ * @cookie:	returned cookie used when freeing the buffer
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread.c]: I'll update the description

**Comment by jenswi-linaro** [comment]
Rebased, squashed and tag applied

**Pull request merged**